src/reducers/vms.js

```diff
@@ -235,31 +156,132 @@ const vms = actionReducer(initialState, {
     return state
   },
 
-  [UPDATE_PAGING_DATA] (state, { payload: { vmsPage, vmsExpectMorePages, poolsPage, poolsExpectMorePages } }) {
-    if (vmsPage) {
-      state = state.set('vmsPage', vmsPage)
-    }
-    state = state.set('vmsExpectMorePages', vmsExpectMorePages)
-
-    if (poolsPage) {
-      state = state.set('poolsPage', poolsPage)
-    }
-    state = state.set('poolsExpectMorePages', poolsExpectMorePages)
-
-    return state
-  },
-
-  [LOGOUT] (state) { // see the config() reducer
+  [C.LOGOUT] (state) { // see the config() reducer
     return state.set('vms', Immutable.fromJS({}))
   },
-  [SET_FILTERS] (state, { payload: { filters } }) { // see the config() reducer
+
+  [C.SET_FILTERS] (state, { payload: { filters } }) { // see the config() reducer
     return state.set('filters', Immutable.fromJS(filters))
   },
-  [SET_VM_SORT] (state, { payload: { sort } }) { // see the config() reducer
+
+  [C.SET_VM_SORT] (state, { payload: { sort } }) { // see the config() reducer
     return state.set('sort', Immutable.fromJS(sort))
   },
 })
 
+/**
+ * Merge an array of VMs in to the current `vms`, optionally retaining VM sub resources.
+ */
+function updateVms (state, updatedVms, keepSubResources) {
+  const updates = {}
+
+  updatedVms.forEach(vm => {
+    const existingVm = state.hasIn(['vms', vm.id]) ? state.getIn(['vms', vm.id]).toJS() : false
+
+    updates[vm.id] = vm
+    updates[vm.id].actionResults = (existingVm && existingVm.actionResults) || {}
+
+    if (existingVm && keepSubResources) {
+      updates[vm.id] = {
+        ...updates[vm.id],
+        ...pullVmSubResources(existingVm),
+      }
+    }
+  })
+
+  state = state.mergeIn(['vms'], Immutable.fromJS(updates))
+
+  const vmsIds = Object.keys(updates)
+  state = state.set('missedVms', state.get('missedVms').subtract(vmsIds))
```

Comment:
`missedVms` is needed for 404/bad id detection with a URL something like `https://fqdn/ovirt-engine/web-ui/vm/ABCD`

Reply:
OK. Why not checking against `vms` directly? i.e. VM  `ABCD` is unlikely to be among `missedVms`

Reply:
Because checking against `vms` isn't enough.  The data flow for opening the app to a VM details:
  - request `https://fqdn/ovirt-engine/web-ui/vm/ABCD`
  - web-ui runs the normal login saga and makes the app ready
  - `CHANGE_PAGE` is fired with the `detailPage` selected
  - `VmDetailsPage` is rendered as the page, but vmId `ABCD` isn't in the `vms` store so it renders `null`
  - `REFRESH_DATA` gets fired with the `vmId` from the URL
  - The fetch for the `vmId` of ABCD fails with a 404
  - Nothing will be added to `store.vms.vms`, and without any other info `VmDetailsPage` will render `null`
  - By adding `ABCD` to the `missedVms` list, `VmDetailsPage` can render a 404 view

So yeah, `missedVms` is needed for that use case.